content/en/blog/_posts/2025-04-21-kubernetes-network-evolution/index.md

```diff
@@ -0,0 +1,71 @@
+
+The first significant step towards directly addressing multi-network requirements within the core Kubernetes project was the introduction of [KEP-3698 (Multi-Network)](https://github.com/kubernetes/enhancements/pull/3700). This Kubernetes Enhancement Proposal aimed to define user stories and requirements for multi-network functionality in Kubernetes. However, early feedback indicated that the KEP was perceived as too abstract and broad, attempting to encompass a wide range of use cases, from simply attaching multiple interfaces to more complex network virtualization scenarios.
+
+A core goal of KEP-3698 was to introduce an API object to Kubernetes describing networks that Pods can attach to and to evolve the current Kubernetes networking model to support multiple networks in a backward-compatible way. The initial focus was on defining what users wanted to achieve with multi-networking, such as connecting Pods to each other in a manner defined by the network object, and exposing basic network interface information for each attachment.
+
+Discussions around the KEP revealed challenges related to integrating multi-networking in existing Kubernetes architecture. For instance, there were questions about how network probes would function on non-default networks (problem that originated a new KEP [https://github.com/kubernetes/enhancements/pull/4558](https://github.com/kubernetes/enhancements/pull/4558)) and the implications for other core functionalities like DNS and network policies in a multi-network environment. KEP-3698 did an important work on clarifying the end user multi-networking requirements. This lack of defined semantics and standard implementations, coupled with the anticipated need to rewrite existing Kubernetes components, ultimately led the community to explore more feasible alternative approaches.
+ 
+### KNI (Kubernetes Networking Interface) KEP: A Reimagined Approach to CNI
+
+As discussions around the limitations of the existing CNI interface for supporting advanced workloads continued, the [KNI (Kubernetes Networking Interface) KEP (KEP-4477)](https://github.com/kubernetes/enhancements/pull/4477) emerged. KNI proposed a more radical reimagining of Kubernetes networking, aiming to reduce the number of layers involved by introducing a gRPC API. The goal was to consolidate networking into a single layer, potentially simplifying troubleshooting and development.
+
+However, the KNI proposal also started an intense debate. Concerns were raised about the necessity of introducing an entirely new API and RPC service in a core component like the Kubelet, instead of improving the existing CNI and CRI (Container Runtime Interface). The potential for significant technical debt and disruption was also a major consideration. Ultimately, the [KNI KEP was closed](https://github.com/kubernetes/enhancements/pull/4477#event-11626847588), signaling a shift in focus towards more incremental and integrated solutions.
```

Comment:
I don't think the closure of the KNI KEP signaled anything other than "this particular KEP is not ever likely to progress to `implementable`". The problem with KNI is that it started out with a PoC of replacing CNI with gRPC, and then turned into a KEP for "fix every problem in every part of k8s networking", without having any real plan for how it was going to do that. It's not, IMO, that we felt it was taking the wrong approach, it's that we never knew what approach it was taking.

Reply:
the closure of signal is when I started to look at the problem in deep 

![image](https://github.com/user-attachments/assets/32220346-2209-4e4b-aef6-20efea8022d8)

and presented in sig network mar 2024 "Things you can do today with the network", to explain that we already have a lot of existing mechanisms to do network things, that was the complain KNI was doing, and we just need to enhance them ... on that investigation, after Kubecon Paris, is when I met NRI folks and DRA and then when we merge everything